src/bind.c

```diff
@@ -122,3 +122,175 @@ static SEXP as_df_row_impl(SEXP x, bool quiet) {
 SEXP vctrs_as_df_row(SEXP x, SEXP quiet) {
   return as_df_row(x, LOGICAL(quiet)[0]);
 }
+
+
+static SEXP as_df_col(SEXP x, SEXP outer);
+static SEXP vec_cbind(SEXP xs, SEXP ptype, SEXP size);
+static SEXP cbind_container_type(SEXP x);
+
+// [[ register(external = TRUE) ]]
+SEXP vctrs_cbind(SEXP call, SEXP op, SEXP args, SEXP env) {
+  args = CDR(args);
+
+  SEXP xs = PROTECT(rlang_env_dots_list(env));
+  SEXP ptype = PROTECT(Rf_eval(CAR(args), env));
+  SEXP size = PROTECT(Rf_eval(CADR(args), env));
+
+  SEXP out = vec_cbind(xs, ptype, size);
+
+  UNPROTECT(3);
+  return out;
+}
+
+static SEXP vec_cbind(SEXP xs, SEXP ptype, SEXP size) {
+  R_len_t n = Rf_length(xs);
+
+  // Find the common container type of inputs
+  SEXP containers = PROTECT(map(xs, &cbind_container_type));
+  ptype = PROTECT(cbind_container_type(ptype));
+
+  SEXP type = PROTECT(vctrs_type_common_impl(containers, ptype));
+  if (type == R_NilValue) {
+    type = new_data_frame(vctrs_shared_empty_list, 0);
+  } else if (!is_data_frame(type)) {
+    type = r_as_data_frame(type);
+  }
+  UNPROTECT(1);
+  PROTECT(type);
+
+
+  R_len_t nrow;
+  if (size == R_NilValue) {
+    nrow = vec_size_common(xs);
```

Comment:
Where do we check for (e.g.) `vec_cbind(NULL, NULL)` ?

Reply:
We skip `NULL` inputs.

Reply:
But what happens if they're all `NULL` and `xs` is empty? I guess we already have a unit test for it so it must work, it's just not super clear how.

Reply:
`nrow` will be 0 in that case. Then we cast an empty list to a data frame (the common prototype).